conn.go

```diff
@@ -641,34 +655,70 @@ func (c *Conn) exec(ctx context.Context, req frameWriter, tracer Tracer) (*frame
 				// this is because the request is still outstanding and we have
 				// been handed another error from another stream which caused the
 				// connection to close.
-				c.releaseStream(stream)
+				c.releaseStream(call.streamID)
 			}
-			return nil, err
+			return err
 		}
+		return nil
 	case <-timeoutCh:
 		close(call.timeout)
 		c.handleTimeout()
-		return nil, ErrTimeoutNoResponse
+		return ErrTimeoutNoResponse
 	case <-ctxDone:
 		close(call.timeout)
-		return nil, ctx.Err()
+		return ctx.Err()
 	case <-c.quit:
-		return nil, ErrConnectionClosed
+		return ErrConnectionClosed
 	}
+}
 
-	// dont release the stream if detect a timeout as another request can reuse
-	// that stream and get a response for the old request, which we have no
-	// easy way of detecting.
-	//
-	// Ensure that the stream is not released if there are potentially outstanding
-	// requests on the stream to prevent nil pointer dereferences in recv().
-	defer c.releaseStream(stream)
+func (c *Conn) sendFrame(ctx context.Context, call *callReq, req frameWriter) error {
+	timeoutCh := c.resetTimeout(call)
+	var ctxDone <-chan struct{}
+	if ctx != nil {
+		ctxDone = ctx.Done()
+	}
 
-	if v := framer.header.version.version(); v != c.version {
-		return nil, NewErrProtocol("unexpected protocol version in response: got %d expected %d", v, c.version)
+	select {
+	case c.frameWriteArgChan <- &frameWriteArg{req, call.framer, call.streamID}:
+		return nil
+	case <-timeoutCh:
+		close(call.timeout)
+		c.handleTimeout()
+		return ErrTimeoutNoResponse
+	case <-ctxDone:
+		close(call.timeout)
+		return ctx.Err()
+	case <-c.quit:
+		return ErrConnectionClosed
 	}
+}
 
-	return framer, nil
+func (c *Conn) resetTimeout(call *callReq) <-chan time.Time {
```

Comment:
it would make more sense to invert this and have it have the signature like

```go
func (c *callReq) resetTimeout(timeout time.Duration) <-chan time.Time
```